transport/cluster.go

```diff
@@ -0,0 +1,288 @@
+		return nil, fmt.Errorf("creating cluster: %w", err)
+	}
+
+	go c.handleEvents()
+	go c.loop()
+	return c, nil
+}
+
+func (c *Cluster) OpenControl() error {
+	// TODO: should we random shuffle known hosts after establishing connection?
+	for _, v := range c.knownHosts {
+		control, err := OpenConn(v, nil, c.cfg)
+		if err == nil {
+			if events, err := control.registerEvents(c.handledEvents); err == nil {
+				c.control = control
+				c.events = events
+				return nil
+			}
+		}
+		if control != nil {
+			control.Close()
+		}
+	}
+	return fmt.Errorf("couldn't open control connection to any known host")
+}
+
+// refreshTopology creates new PeerMap filled with the result of both localQuery and peerQuery.
+// The old map is replaced with the new one atomically to prevent dirty reads.
+func (c *Cluster) refreshTopology() error {
+	rows, err := c.getAllNodesInfo()
+	if err != nil {
+		return fmt.Errorf("refreh topology: %w", err)
+	}
+	// If node is present in both maps we can update already created node
+	// instead of creating new one from the scratch.
+	old := c.GetPeers()
+	m := PeerMap{}
+	for _, v := range rows {
+		addr := net.IP(v[nodeAddr]).String()
+
+		if node, ok := old[addr]; ok {
+			node.tokens = v[nodeTokens]
+			m[addr] = node
+		} else {
+			n := &Node{
+				addr:       addr,
+				datacenter: string(v[nodeDC]),
+				rack:       string(v[nodeRack]),
+				tokens:     v[nodeTokens],
+			}
+			if pool, err := NewConnPool(appendDefaultPort(addr), c.cfg); err != nil {
+				n.status = atomic.NewBool(statusDown)
+			} else {
+				n.status = atomic.NewBool(statusUP)
+				n.pool = pool
+			}
+			m[addr] = n
+		}
+	}
+
+	for k, v := range old {
+		if _, ok := m[k]; !ok {
+			v.pool.Close()
+		}
+	}
+	c.SetPeers(m)
+	return nil
+}
+
+var (
+	peerQuery = Statement{
+		Content:     "SELECT peer, data_center, rack, tokens FROM system.peers",
+		Consistency: frame.ONE,
+	}
+
+	localQuery = Statement{
+		Content:     "SELECT rpc_address, data_center, rack, tokens FROM system.local",
+		Consistency: frame.ONE,
+	}
+)
+
+// Indexes of columns returned from refresh topology queries.
+const (
+	nodeAddr   = 0
+	nodeDC     = 1
+	nodeRack   = 2
+	nodeTokens = 3
+)
+
+func (c *Cluster) getAllNodesInfo() ([]frame.Row, error) {
+	peerRes, err := c.control.Query(peerQuery, nil)
+	if err != nil {
+		return nil, fmt.Errorf("discovering peer topology: %w", err)
+	}
+
+	localRes, err := c.control.Query(localQuery, nil)
+	if err != nil {
+		return nil, fmt.Errorf("discovering local topology: %w", err)
+	}
+
+	return append(peerRes.Rows, localRes.Rows[0]), nil
+}
+
+func (c *Cluster) GetPeers() PeerMap {
+	return c.peers.Load().(PeerMap)
+}
+
+func (c *Cluster) SetPeers(m PeerMap) {
+	c.peers.Store(m)
+}
+
+const defaultPort = 9042
+
+func appendDefaultPort(addr string) string {
+	return addr + ":" + strconv.Itoa(defaultPort)
+}
+
+const eventChanSize = 32
+
+func (c *Cluster) handleEvents() {
+	c.events = make(responseHandler, eventChanSize)
```

Comment:
I don't understand we would me make channel here and also in registerEvents?

Reply:
It's a bug, thanks for catching it.